File: src/main/java/com/bdev/hogwarts_api/service/student_attendance/StudentAttendanceServiceImpl.java
package com.bdev.hogwarts_api.service.student_attendance;

import com.bdev.hogwarts_api.dao.StudentAttendanceDao;
import com.bdev.hogwarts_api.data.converter.student_attendance.StudentAttendanceDtoConverter;
import com.bdev.hogwarts_api.data.converter.student_attendance.StudentAttendanceModelConverter;
import com.bdev.hogwarts_api.data.dto.student.StudentAttendance;
import com.bdev.hogwarts_api.service.student.StudentService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

import static java.lang.String.format;
import static java.util.stream.Collectors.toList;

@Service
public class StudentAttendanceServiceImpl implements StudentAttendanceService {
    @Autowired
    private StudentAttendanceDao studentAttendanceDao;

    @Autowired
    private StudentService studentService;

    @Override
    public boolean exists(long attendanceId) {
        return studentAttendanceDao.exists(attendanceId);
    }

    @Override
    public List<StudentAttendance> getAllAttendances() {
        return studentAttendanceDao
                .findAll()
                .stream()
                .map(StudentAttendanceModelConverter::convert)
                .collect(toList());
    }

    @Override
    public List<StudentAttendance> getAttendances(long studentId) {
        if (!studentService.exists(studentId)) {
            throw new RuntimeException(format("Student with id '%d' does not exist", studentId));
        }

        return studentAttendanceDao
                .getAllByStudentId(studentId)
                .stream()
                .map(StudentAttendanceModelConverter::convert)
                .collect(toList());
    }

    @Override
    public long addAttendance(StudentAttendance attendance) {
        if (!studentService.exists(attendance.getStudentId())) {
            throw new RuntimeException(format("Student with id '%d' does not exist", attendance.getStudentId()));
        }

        return studentAttendanceDao.save(StudentAttendanceDtoConverter.convert(attendance)).getId();
    }

    @Override
    public void deleteAttendance(long attendanceId) {
        if (!studentAttendanceDao.exists(attendanceId)) {
            throw new RuntimeException(format("Student attendance with id '%d' does not exist", attendanceId));
        }

        studentAttendanceDao.delete(attendanceId);
    }
}

File: src/main/java/com/bdev/hogwarts_api/service/student_attendance/StudentAttendanceServiceImpl.kt
package com.bdev.hogwarts_api.service.student_attendance

import com.bdev.hogwarts_api.dao.StudentAttendanceDao
import com.bdev.hogwarts_api.data.converter.student_attendance.StudentAttendanceDtoConverter
import com.bdev.hogwarts_api.data.converter.student_attendance.StudentAttendanceModelConverter
import com.bdev.hogwarts_api.data.dto.student.StudentAttendance
import com.bdev.hogwarts_api.data.dto.student.StudentAttendanceId
import com.bdev.hogwarts_api.data.model.student_attendance.StudentAttendanceModelId
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class StudentAttendanceServiceImpl @Autowired constructor(
        private val studentAttendanceDao: StudentAttendanceDao
): StudentAttendanceService {
    override fun getAllAttendances(): List<StudentAttendance> {
        return studentAttendanceDao
                .findAll()
                .map { StudentAttendanceModelConverter.convert(it) }
    }

    override fun getStudentAttendances(studentLogin: String): List<StudentAttendance> {
        return studentAttendanceDao
                .getAllByIdStudentLogin(studentLogin)
                .map { StudentAttendanceModelConverter.convert(it) }
    }

    override fun addAttendance(attendance: StudentAttendance) {
        studentAttendanceDao.save(StudentAttendanceDtoConverter.convert(attendance))
    }

    override fun deleteAttendance(attendanceId: StudentAttendanceId) {
        studentAttendanceDao.delete(
                StudentAttendanceModelId(
                        studentLogin = attendanceId.studentLogin,
                        startTime = attendanceId.startTime,
                        finishTime = attendanceId.finishTime
                )
        )
    }
}
